src/core/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -5,78 +5,123 @@
 import Glean from "@mozilla/glean/webext";
 import { custom } from "./generated/pings";
 import { config } from "../../../config";
+import {
+  modelLoadTime,
+  translationTime,
+  wordsPerSecond,
+} from "./generated/performance";
+import { fromLang, toLang } from "./generated/metadata";
+import {
+  displayed,
+  changeLang,
+  closed,
+  neverTranslateSite,
+  translate,
+  neverTranslateLang,
+  notNow,
+} from "./generated/infobar";
 
+/**
+ * This class contains general telemetry initialization and helper code and synchronous telemetry-recording functions.
+ *
+ * Synchronous methods here is important, since it is the only way to guarantee that multiple Glean API calls are
+ * executed sequentially and not interleaved with other asynchronous Telemetry recording.
+ * For more information, see: https://github.com/mozilla-extensions/bergamot-browser-extension/pull/76#discussion_r602128568
+ */
 export class Telemetry {
-  private static _instance = null;
-  private _metricsToSubmit;
-
   constructor() {
     const appId = config.telemetryAppId;
     Glean.initialize(appId, true, {
       debug: { logPings: config.telemetryDebugMode },
     });
-    this._metricsToSubmit = 0;
     console.info(
       `Telemetry: initialization completed with application ID ${appId}.`,
     );
-    // Disabled telemetry until opt-out-preference is respected (https://github.com/mozilla-extensions/bergamot-browser-extension/issues/51)
-    this.setUploadEnabled(false);
-    console.info(`Note: Telemetry upload is disabled in this build`);
   }
 
-  /**
-   * Provides access to Telemetry singleton.
-   */
-  public static get global(): Telemetry {
-    if (Telemetry._instance === null) {
-      Telemetry._instance = new Telemetry();
-    }
-    return Telemetry._instance;
+  public onInfoBarDisplayed(tabId: number) {
+    displayed.record();
+    this.submit();
   }
 
-  /**
-   * Enables uploading Glean metrics to telemetry server.
-   * @param val bool
-   */
-  public setUploadEnabled(val: boolean) {
-    Glean.setUploadEnabled(val);
+  public onSelectTranslateFrom(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onSelectTranslateTo(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onInfoBarClosed(tabId: number) {
+    closed.record();
+    this.submit();
+  }
+
+  public onNeverTranslateSelectedLanguage(tabId: number) {
+    neverTranslateLang.record();
+    this.submit();
+  }
+
+  public onNeverTranslateThisSite(tabId: number) {
```

Comment:
As you mention, we'll be able to find the language pair by joining subsequent events with the Infobar display event from the same user session. Is this enough for this PR, or should I also add from/to metadata to all infobar events?

Reply:
joining is always expensive, we should write all relevant metadata for all metrics. Joining for SelectTranslateTo is an edge case to investigate some specific issues and not overload metrics with extra metadata for this specific case. So yes, please add metadata for infobar events. It will be also applicable when we add telemetry client id.